File: internal/crypto/password.go
package crypto

import (
	"crypto/rand"
	"errors"
	"io"

	"golang.org/x/crypto/argon2"
	"golang.org/x/crypto/blake2b"
)

// Password-protected reveals (docs/crypto-spec.md section 4.1).
//
// A reveal link normally carries the whole decryption key, so whoever holds
// the link can open it. When the human has set a reveal password, the agent
// mixes a password-derived key into the link key: the link alone is not
// enough, and the password alone is not enough. The parameters below are
// libsodium's crypto_pwhash interactive limits with Argon2id 1.3, so the
// browser (libsodium.js), PyNaCl, and Go derive the same bytes.

const (
	// SaltSize is the size of the per-reveal password salt carried in the link.
	SaltSize = 16
	// PasswordOpsLimit is the Argon2id time cost (crypto_pwhash_OPSLIMIT_INTERACTIVE).
	PasswordOpsLimit = 2
	// PasswordMemLimitKiB is the Argon2id memory cost in KiB (crypto_pwhash_MEMLIMIT_INTERACTIVE, 64 MiB).
	PasswordMemLimitKiB = 64 * 1024
	// MinPasswordLen is the shortest reveal password the CLI accepts.
	MinPasswordLen = 8
	// passwordDomain separates this derivation from any other use of BLAKE2b.
	passwordDomain = "burndrop/reveal-password/v1"
)

// ErrPassword is returned for an empty password or a salt of the wrong size.
var ErrPassword = errors.New("crypto: password or salt invalid")

// NewSalt returns a random salt for one password-protected reveal. random
// may be nil, in which case crypto/rand is used.
func NewSalt(random io.Reader) ([]byte, error) {
	if random == nil {
		random = rand.Reader
	}
	salt := make([]byte, SaltSize)
	if _, err := io.ReadFull(random, salt); err != nil {
		return nil, err
	}
	return salt, nil
}

// PasswordKey derives 32 bytes from a password and a salt with Argon2id
// (time 2, memory 64 MiB, one lane), the libsodium interactive parameters.
// The password is used as typed, in UTF-8, with no normalization.
func PasswordKey(password, salt []byte) (*[KeySize]byte, error) {
	if len(password) == 0 || len(salt) != SaltSize {
		return nil, ErrPassword
	}
	out := argon2.IDKey(password, salt, PasswordOpsLimit, PasswordMemLimitKiB, 1, KeySize)
	key := new([KeySize]byte)
	copy(key[:], out)
	Zero(out)
	return key, nil
}

// RevealKeyWithPassword combines the key carried in the link with the
// password-derived key: BLAKE2b-256(domain || linkKey || passwordKey). The
// result is the XChaCha20-Poly1305 key of the reveal.
func RevealKeyWithPassword(linkKey *[KeySize]byte, password, salt []byte) (*[KeySize]byte, error) {
	pw, err := PasswordKey(password, salt)
	if err != nil {
		return nil, err
	}
	defer ZeroKey(pw)
	return combineKeys(linkKey, pw), nil
}

func combineKeys(linkKey, passwordKey *[KeySize]byte) *[KeySize]byte {
	msg := make([]byte, 0, len(passwordDomain)+2*KeySize)
	msg = append(msg, passwordDomain...)
	msg = append(msg, linkKey[:]...)
	msg = append(msg, passwordKey[:]...)
	sum := blake2b.Sum256(msg)
	Zero(msg)
	key := new([KeySize]byte)
	copy(key[:], sum[:])
	Zero(sum[:])
	return key
}

File: sdk/python/src/burndrop/password.py
"""Password-protected reveals (crypto specification section 4.1).

A reveal link normally carries the whole decryption key. When the human has
set a reveal password, the agent mixes a password-derived key into the link
key: the link alone is not enough, and the password alone is not enough.
The Argon2id parameters are libsodium's interactive limits, so PyNaCl, the
browser (libsodium.js), and Go derive the same bytes.
"""

from __future__ import annotations

import hashlib
import os

import nacl.pwhash

from .crypto import KEY_SIZE, CryptoError, LengthError

SALT_SIZE = 16
"""Size of the per-reveal salt carried in the link."""

PASSWORD_OPSLIMIT = 2
"""Argon2id time cost (crypto_pwhash_OPSLIMIT_INTERACTIVE)."""

PASSWORD_MEMLIMIT = 64 * 1024 * 1024
"""Argon2id memory cost in bytes (crypto_pwhash_MEMLIMIT_INTERACTIVE, 64 MiB)."""

_DOMAIN = b"burndrop/reveal-password/v1"


def new_salt() -> bytes:
    """Return a random salt for one password-protected reveal."""
    return os.urandom(SALT_SIZE)


def password_key(password: str | bytes, salt: bytes) -> bytes:
    """Derive 32 bytes from a password with Argon2id 1.3 (time 2, memory 64 MiB, one lane).

    The password is used as typed, in UTF-8, with no normalization.
    """
    raw = password.encode("utf-8") if isinstance(password, str) else bytes(password)
    if not raw:
        raise CryptoError("password must not be empty")
    if len(salt) != SALT_SIZE:
        raise LengthError(f"salt must be {SALT_SIZE} bytes")
    derived = nacl.pwhash.argon2id.kdf(
        KEY_SIZE, raw, bytes(salt), opslimit=PASSWORD_OPSLIMIT, memlimit=PASSWORD_MEMLIMIT
    )
    return bytes(derived)


def reveal_key_with_password(link_key: bytes, password: str | bytes, salt: bytes) -> bytes:
    """Return BLAKE2b-256(domain || link_key || password_key).

    That is the XChaCha20-Poly1305 key of a password-protected reveal.
    """
    if len(link_key) != KEY_SIZE:
        raise LengthError(f"key must be {KEY_SIZE} bytes")
    material = _DOMAIN + bytes(link_key) + password_key(password, salt)
    return hashlib.blake2b(material, digest_size=KEY_SIZE).digest()
